lib/web_ui/lib/src/engine/text/font_collection.dart

```diff
@@ -96,40 +96,78 @@ class FontCollection {
 
-  factory _FontManager() {
+  factory FontManager() {
     if (supportsFontLoadingApi) {
-      return _FontManager._();
+      return FontManager._();
     } else {
       return _PolyfillFontManager();
     }
   }
 
-  _FontManager._();
-
+  FontManager._();
+
+  /// Registers assets to Flutter Web Engine.
+  ///
+  /// Browsers and browesers versions differ siginificantly on how a valid font
+  /// family name should be formatted. Notable issues are:
+  /// Safari12 and Firefox crash if you create a [html.FontFace] with a font
+  /// family that is not correct CSS syntax. Font family names accepted on these
+  /// browsers, when wrapped it in quotes.
+  /// Additionally, for Safari12 to work [html.FontFace] name should be loaded
+  /// correctly on the first try.
+  /// A font in Chrome chrashes if a [html.FontFace] is loaded only with quotes.
+  /// Unlike Safari12 if a valid version is loaded afterwards it will show
+  /// that fonts normally.
+  /// In Safari 13 the [html.FontFace] should be loaded with unquoted family
+  /// names.
+  /// In order to avoid all these browser compatibility issues this method;
+  /// detects the family names that might cause a conflict, loads it with
+  /// quotes and loads it again without quotes. For all the other family names
+  /// [html.FontFace] is loaded only once.
+  /// See: https://developer.mozilla.org/en-US/docs/Web/CSS/font-family#Valid_family_names
+  /// See: https://drafts.csswg.org/css-fonts-3/#font-family-prop
   void registerAsset(
     String family,
     String asset,
     Map<String, String> descriptors,
   ) {
-    // Safari and Firefox crash if you create a [html.FontFace] with a font
-    // family that is not correct CSS syntax. To ensure the font family is
-    // accepted on these browsers, wrap it in quotes.
-    // See: https://drafts.csswg.org/css-fonts-3/#font-family-prop
-    if (browserEngine == BrowserEngine.webkit || browserEngine == BrowserEngine.firefox) {
-      family = "'$family'";
+    final String familyNameInQuotes = "'$family'";
+    // Regular expression to detect punctuations. For example font family
+    // 'Ahem!' falls into this category.
+    final RegExp punctuations = RegExp(r"[.,:;!`\/#\$\%\^&~\*=\-_(){}]");
+    // Regular expression to detect tokens starting with a digit.
+    // For example font family 'Goudy Bookletter 1911' falls into this
+    // category.
+    final RegExp startWithDigit = RegExp(r"\b\d");
+    // Fonts names when a package dependency is added has '/' in the family
+    // names such as 'package/material_design_icons_flutter/...'
+    if (family.contains('/') ||
+        punctuations.hasMatch(family) ||
+        startWithDigit.hasMatch(family)) {
+      // Load a font family name with special chracters once here wrapped in
+      // quotes.
+      _loadFontFace(familyNameInQuotes, asset, descriptors);
     }
+    // Load all font fonts, without quoted family names.
+    _loadFontFace(family, asset, descriptors);
```

Comment:
I'll delegate this to @hterkelsen. Is there risk that users will see errors in the browser from us attempting to load one of the fonts incorrectly? Perhaps we could try the best known way to load fonts for each browser/version we know about, and fallback to this strategy for browser/version we don't know?

Reply:
No, there is no risk. I did tests in every browser.

I am not in favor of that approach. If the best way we know about loads an invalid font in Safari, we won't be able to load fonts before we clear(html.document.fonts.clear). We should do what works Safari 12 and then presume. 

As I mentioned in yesterday's meeting I worked on 5 different solutions this week.

Reply:
SGTM